### ISSUE TITLE
cdc: make it possible to safely add an option without adding a cluster version

### ISSUE THREAD

**Issue by HonoreDB** [cockroachdb/cockroach#83126]
Currently, if we added support for the option `WITH redact_sensitive_data`, we'd have two choices:

1. Add a new cluster version, SupportsRedactedChangefeed, and add a version gate on it to changefeedbase VersionGateOptions. This makes the feature unbackportable, as cluster versions automatically imply backwards incompatibility.
2. Do nothing. Then, if the changefeed is started on a mixed-version cluster, changefeed aggregator nodes from a version before the option will emit unredacted data.

It should instead be easy to add a WITH option such that changefeeds can't be created until all nodes support it, and will error out if any node is downgraded to no longer recognize it, without making the change un-backportable.

Strawman implementation: Create a generated file called `utils/lightweight_versions.go` that maps a namespace and arbitrary string to a cluster version. The command to make a release parses all go files and looks for all calls to `lightweight_versions.VersionFor(str)` where str is not yet in the map, and adds it to the map pointing it to the release just being created. On creating a changefeed, we make a best-effort attempt to check that the minimum version in the cluster is greater than the VersionFor every option. On starting up a changefeed aggregator or changefeed frontier, we validate that all options are recognized and fail the whole feed if they're not.

**Comment by dt** [comment]
We discussed this on slack at length, but I'm skeptical that we want a generalized framework for "is the cluster running version X?" that is used to decide if it is safe to use new behaviors in version X other than the current cluster version framework. The reason why the current system does not allow for backportability is the same property that gives the system its safety around "If I checked and found X, it is safe to assume X";  a weaker system, like the strawman above could see an old node join the cluster moments after the check said X, so you'd still need to plumb specific checks everywhere, in each rpc, in each distsql proc, etc to ensure that the remote processing that must happen on version X or above is indeed on X or above; a single check at the root will not suffice with a system weaker than cluster version. And if you need to plumb a required version or proxy for required version explicitly in all your requests, proc specs, persisted job records, etc anyway, then you may as well just populate that with a specific minimum version. So I'm not sure that in the end a "lightweight" cluster version actually buys us anything, but could encourage misuse (checking and assuming) where a proper cluster version was actually needed.

The real answer here is to not backport features that need version gates in the first place, which would be easier to abide by if we released major versions more often.

**Comment by HonoreDB** [comment]
>then you may as well just populate that with a specific minimum version

I guess this part is where I get off the train. In the changefeed options case, anyway, the plumbing is already there--anything that needs to be able to honor an option can see that option. But mapping new options to specific minimum versions is still annoying. Even when you can use real version gates it's a bit annoying and people don't always do it (although we could make that part more automatic).

Agree that making it look general invites misuse, though.

**Comment by ajwerner** [comment]
>  But mapping new options to specific minimum versions is still annoying.

In the grand scheme of things we do as a part of this job, this hardly seems that onerous. It feels like the answer is clear: make a framework that enforces you to map new options to a version so that it's obvious and required.

**Comment by HonoreDB** [comment]
>make a framework that enforces you to map new options to a version so that it's obvious and required.

Yup! And if I'm doing that I might as well make the mapping happen automatically.
